### PR TITLE
Ray 1.4.1

### PR DESCRIPTION
**Pull request by krfricke**
Build passes locally

Closes #20

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

cc @vnlitvinov

### CODE REVIEW

**Comment by h-vetinari** [comment]
@conda-forge-admin, please rerender

File: recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch
Comment (vnlitvinov) [inline recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch:34]: this is the exact reason why we have a patch removing the requirements as we do - please update `meta.yaml` with this requirement!

File: recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch
Comment (vnlitvinov) [inline recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch:52]: this implies an update to `meta.yaml` is needed, too

File: recipe/patches/0013-Add-bazel-linkopts-libs.patch
Comment (vnlitvinov) [inline recipe/patches/0013-Add-bazel-linkopts-libs.patch:19]: why do we need this?

**Review by vnlitvinov — CHANGES_REQUESTED** [review]
Thanks @krfricke for taking this up!
Please update requirements in `meta.yaml`, and we should be good to go when CI passes.

File: recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch
Comment (h-vetinari) [reply recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch:34]: Well, this requirement has to do with which version of numpy first supported python 3.9. I don't think we need to amend the current `numpy >=1.16`, because conda obviously also has no builds for 3.9 before 1.19

File: recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch
Comment (h-vetinari) [reply recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch:52]: Similarly here; strictly speaking you're correct, but we already only pin `cython >=0.29`, which is more than enough (also 0.29.15 is ancient)

**Review by h-vetinari — COMMENTED** [review]
Thanks for taking this up!

The rerender I triggered was so that you actually become a maintainer when this PR is merged. 🙃

File: recipe/patches/0013-Add-bazel-linkopts-libs.patch
Comment (krfricke) [reply recipe/patches/0013-Add-bazel-linkopts-libs.patch:19]: Local build wouldn't pass otherwise (getting linker errors in grpc/protobuf builds)

**Comment by krfricke** [comment]
Thanks for your help reviewing this! Updated the dependencies in the meta.yaml

**Comment by h-vetinari** [comment]
> Thanks for your help reviewing this! Updated the dependencies in the meta.yaml

Did you see my comments? This is a misspecification (cython is debatable, but numpy certainly). I'd ask you to revert (or rebase out) that commit.

**Comment by krfricke** [comment]
There are CI failures that seem to be unrelated:

```
+ conda install --yes --quiet conda-forge-ci-setup=3 conda-build pip -c conda-forge
Collecting package metadata (current_repodata.json): ...working... 
CondaHTTPError: HTTP 000 CONNECTION FAILED for url <https://conda.anaconda.org/conda-forge/noarch/current_repodata.json>
Elapsed: -

An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.
'https://conda.anaconda.org/conda-forge/noarch'
```

can we re-trigger those tests?

**Comment by krfricke** [comment]
In this commit: https://github.com/conda-forge/ray-packages-feedstock/pull/28/commits/fbb74dc4d7a57010680475520e384f44558747ef
I updated the meta to reflect the dependencies from the setup.py - I thought that was what you meant, or did I misunderstand?

**Comment by krfricke** [comment]
Ah sorry I became confused by the comments. Happy to revert the commit

File: recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch
Comment (vnlitvinov) [reply recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch:52]: If so, we should pin to `>=0.29.15` anyway (or maybe something more recent) just to be 100% sure (instead of 99.9% as is now 😆)

**Comment by vnlitvinov** [comment]
@h-vetinari do you know what's going on? seems like something very low in the dependency chain had updated `glibc` (I wonder what it was), but the image used here does not have new enough glibc...

**Comment by h-vetinari** [comment]
Can you try adding `- sysroot_linux-64 2.17  # [linux64]` to the build requirement of the failing package (presumably `ray-core`)?

**Comment by h-vetinari** [comment]
Ah, and we probably will need something like
```
cdt_name:  # [linux]
  - cos7   # [linux]
```

in `conda_build_config.yaml` (needs to be created; then rerendered)

**Comment by vnlitvinov** [comment]
I am still wary about moving to `cos7` as I saw compatibility issues with other packages - is it now the default version in conda-forge?

**Comment by krfricke** [comment]
I added sysroot linux - let me know about cos7

**Comment by h-vetinari** [comment]
> 
> 
> I am still wary about moving to `cos7` as I saw compatibility issues with other packages - is it now the default version in conda-forge?

CentOS 6 is EOL since more than half a year. conda-forge is still discussing about migrating (https://github.com/conda-forge/conda-forge.github.io/issues/1436) but most people are ready to do it.

The compatibility issues should be simple (and mostly in conda-forge CI) - you cannot be on cos6 and also depend on a package built against glibc 2.17.

By adding the `nodejs`-dependency, we won't be able to avoid it here: https://github.com/conda-forge/nodejs-feedstock/blob/master/recipe/meta.yaml#L33

**Comment by h-vetinari** [comment]
> I added sysroot linux - let me know about cos7

It's probably gonna fail because the default image does not have a new enough glibc to satisfy that constraint.

**Comment by krfricke** [comment]
@conda-forge-admin, please rerender

**Comment by krfricke** [comment]
```
subspace that is not fully implemented. To be clear:
.. we did not find ['cos7'] from {'cdt_name': ['cos7']} in cdt_name:['cos6', 'cos6']
```

Not familiar with this config, any pointers? I'll try to look this up in the meantime

**Comment by h-vetinari** [comment]
> Not familiar with this config, any pointers? I'll try to look this up in the meantime

Sorry, I always forget the details; let me fix it.

**Comment by h-vetinari** [comment]
Do you mind if I push into your PR?

**Comment by krfricke** [comment]
Not at all, please go ahead!

**Review by h-vetinari — APPROVED** [review]
Linux passes, hopefully windows will too; LGTM (though I realized too late that the build number should ideally still be reset to 0)

File: recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch
Comment (h-vetinari) [inline recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch:3]: Sidenote: we try to keep the attribution to the original authors when rebasing patches; less for the glory and more for knowing how to ask about it ;-)

Comment (krfricke) [reply recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch:3]: Ah sorry! I just `git am`'d everything and created new patches with `git format-patch` which seems to have lost the original authors. I'll try to sort this out

Comment (krfricke) [reply recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch:3]: Ah so this was only in this patch, and because I actually didn't `git am` it after all (because of conflicts) but manually did the required changes again and committed with the same commit message. What is the best way to go about this next time?

Comment (h-vetinari) [reply recipe/patches/0010-Empty-install-requirements-installed-by-conda-build.patch:3]: > What is the best way to go about this next time?

For me personally, I keep a branch in a local copy of the upstream repo that I rebase onto new versions, and if a conflict occurs, I try to use `git commit -m "<original_message>" --author="<original_author>" --date="<original_date>"` after resolving possible conflicts (if none exist, it'll be fine already).

File: recipe/conda_build_config.yaml
Comment (vnlitvinov) [inline recipe/conda_build_config.yaml:4]: so there is no way to stay on `cos6`?

File: recipe/meta.yaml
Comment (vnlitvinov) [inline recipe/meta.yaml:74]: if we update the base to `cos7` do we need to manually add this?..

File: recipe/meta.yaml
Comment (vnlitvinov) [inline recipe/meta.yaml:309]: welcome on board! 🙃

**Review by vnlitvinov — APPROVED** [review]
Mostly good, so stamping approval.
Would like to see my questions answered, though.

File: recipe/meta.yaml
Comment (krfricke) [reply recipe/meta.yaml:309]: Thanks 😄

**Comment by krfricke** [comment]
FYI I prepared the update for Ray 1.5.0 based on this revision and will open the PR once this has been merged

File: recipe/conda_build_config.yaml
Comment (h-vetinari) [reply recipe/conda_build_config.yaml:4]: No, we cannot, see [this](https://github.com/conda-forge/ray-packages-feedstock/pull/28#issuecomment-894279918) comment.

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:74]: It's necessary so that this package gets the correct run-export (so that packages depending on ray fail to build if they don't have cos7 themselves), otherwise we might get crashes with the ABI.

File: recipe/meta.yaml
Comment (h-vetinari) [reply recipe/meta.yaml:309]: Agreed, already mentioned that on the last PR - happy to have you on board!

**Review by h-vetinari — COMMENTED** [review]
Let's put this in 🙃

File: recipe/conda_build_config.yaml
Comment (vnlitvinov) [reply recipe/conda_build_config.yaml:4]: But we can use older `nodejs`, e.g. one from `14.x` version part, which is still in "active" phase according to their schedule and works on `cos6` :)

**Comment by vnlitvinov** [comment]
@h-vetinari I've removed "automerge" label because I want to finish the `cos6` vs `cos7` discussion first.

**Comment by h-vetinari** [comment]
> 
> 
> @h-vetinari I've removed "automerge" label because I want to finish the `cos6` vs `cos7` discussion first.

Unless you can remove the `nodejs` dependency, there is no wiggle room on this.

**Comment by h-vetinari** [comment]
I don't understand the problem though - cos6 is EOL for almost a year, glibc 2.12 is a decade old. There's a vanishingly small fraction of users affected by moving to cos7.

The only pain is having to build all packages dependent on ray with cos7 as well, but that's manageable IMO. Also conda-forge will move to cos7 in the foreseeable future (see link above), it just takes a bit longer when it's on such a scale.

File: recipe/conda_build_config.yaml
Comment (h-vetinari) [reply recipe/conda_build_config.yaml:4]: Ah, this comment didn't show up in the main thread for some reason.

I'm strongly -1 against pinning to old versions for the sake of EOL distros.

Comment (h-vetinari) [reply recipe/conda_build_config.yaml:4]: If you feel so strongly about cos6-support, we can add a branch where nodejs is pinned to 14.x, but master must move on. (though this would need some care with the build numbers; e.g. build numbers on master start from 100)

**Comment by h-vetinari** [comment]
@vnlitvinov 
I'd like to merge this and make our way towards 1.5.1. If you want to have a branch for cos6 (pinning old nodejs), we can do that, but should then adapt the build number scheme before merging (the higher ones being reserved for master branch). Obviously, I'd prefer to avoid that hassle and just require cos7, because - as laid out above - I really don't see the issue with doing that.

**Comment by vnlitvinov** [comment]
Okay, let's be bold and just move to `cos7` - on our way to the shiny future! Thanks @krfricke again for your efforts and for taking this up!

**Pull request merged**
